### PR TITLE
Initial unit tests to DensityEstimator class

### PR DESCRIPTION
**Pull request by williamfgc**
## Proposed changes

Start setting up simple unit tests for the DensityEstimator class
Current coverage comes from deterministic (functional) tests, the goal is to add more coverage from unit tests.
Part of #3411 follows PR #4371

## What type(s) of changes does this code introduce?

- Refactoring (no functional changes, no api changes)
- Testing changes (e.g. new unit/integration/performance tests)

### Does this introduce a breaking change?

- No

## What systems has this change been tested on?
Ubuntu 22.04 

## Checklist

_Update the following with a yes where the items apply. If you're unsure about any of them, don't hesitate to ask.  This is
simply a reminder of what we are going to look for before merging your code._

- Yes. This PR is up to date with current the current state of 'develop'
- Yes. Code added or changed in the PR has been clang-formatted
- Yes. This PR adds tests to cover any new code, or to catch a bug that is being fixed
- No. Documentation has been added (if appropriate)

### CODE REVIEW

File: src/QMCHamiltonians/tests/test_density_estimator.cpp [hunk 1]
```diff
@@ -0,0 +1,73 @@
+#include "catch.hpp"
```
[the remark this answers is not in the document]
Comment (williamfgc) [reply src/QMCHamiltonians/tests/test_density_estimator.cpp:11]: Good catch, thanks!

File: src/QMCHamiltonians/tests/test_density_estimator.cpp [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+  elec.resetGroups();
+  elec.update();
+
+  DensityEstimator density_estimator(elec);
+
+  ParticleSet::Walker_t dummy = ParticleSet::Walker_t(1);
+  density_estimator.setHistories(dummy);
+
+  const double val = density_estimator.evaluate(elec);
+  CHECK(val == Approx(0.0));
```
Comment (ye-luo) [inline src/QMCHamiltonians/tests/test_density_estimator.cpp:57]: Better to have non-zero value to check.

File: src/QMCHamiltonians/tests/test_density_estimator.cpp [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+  ParticleSet elec(simulation_cell);
+
+  elec.setName("elec");
+  elec.create({2});
+  elec.R[0][0] = 0.0;
+  elec.R[0][1] = 1.0;
+  elec.R[0][2] = 0.0;
+  elec.R[1][0] = 0.4;
+  elec.R[1][1] = 0.3;
+  elec.R[1][2] = 0.0;
```
Comment (ye-luo) [inline src/QMCHamiltonians/tests/test_density_estimator.cpp:36]: use compact lines `elec.R[0] = {0.0, 1.0, 0.0};`

File: src/QMCHamiltonians/tests/test_density_estimator.cpp (hunk 2 above)
Comment (williamfgc) [reply src/QMCHamiltonians/tests/test_density_estimator.cpp:57]: Yes, it's always returning zero at completion, that's the motivation for checking.

**Comment by williamfgc** [comment]
Test this please

**Comment by markdewing** [comment]
Test this please

**Pull request merged**